File: android/app/src/main/java/com/tisza/esemenynaptar/database/EventDatabase.java
package com.tisza.esemenynaptar.database;

import android.content.*;
import android.os.*;
import androidx.annotation.*;
import androidx.room.*;
import androidx.sqlite.db.*;

import java.io.*;

@Database(entities = {Event.class}, version = 1, exportSchema = false)
@TypeConverters({CategoryConverter.class})
public abstract class EventDatabase extends RoomDatabase
{
	private static final String NAME = "event.db";
	private static final String NAME_SRC = "event_src.db";
	private static EventDatabase instance;
	private static EventDatabase tempDatabase;

	public abstract EventDao eventDao();

	public static synchronized void init(Context context, Runnable onReady)
	{
		tempDatabase = Room.databaseBuilder(context.getApplicationContext(), EventDatabase.class, NAME).addCallback(new Callback()
		{
			@Override
			public void onOpen(@NonNull SupportSQLiteDatabase db)
			{
				db.execSQL("ATTACH DATABASE ? AS event_src;", new Object[]{copyAssetDatabaseToData(context)});
				db.execSQL("DROP TABLE IF EXISTS event_old;");
				db.execSQL("CREATE TABLE event_old AS SELECT * FROM event;");
				db.execSQL("DELETE FROM event;");
				db.execSQL("INSERT INTO event SELECT event_src.event.year, event_src.event.month, event_src.event.day, event_src.event.category, event_src.event.text, ifnull(event_old.isLiked, 0) "
						+ "FROM event_src.event LEFT JOIN event_old "
						+ "ON event_src.event.year = event_old.year AND event_src.event.month = event_old.month AND event_src.event.day = event_old.day AND event_src.event.category = event_old.category;");
				db.execSQL("DROP TABLE IF EXISTS event_old;");

				new Handler(context.getMainLooper()).post(() ->
				{
					tempDatabase.close();
					instance = Room.databaseBuilder(context.getApplicationContext(), EventDatabase.class, NAME).build();
					onReady.run();
				});
				super.onOpen(db);
			}
		}).build();

		new AsyncTask<Void, Void, Void>()
		{
			@Override
			protected Void doInBackground(Void... voids)
			{
				try
				{
					tempDatabase.query("SELECT 0", null);
				}
				catch (Exception e)
				{
					e.printStackTrace();
				}
				return null;
			}
		}.execute();
	}

	private static String copyAssetDatabaseToData(Context context)
	{
		File dbPath = context.getDatabasePath(NAME_SRC);
		dbPath.getParentFile().mkdirs();

		try
		{
			final InputStream input = context.getAssets().open(NAME);
			final OutputStream output = new FileOutputStream(dbPath);

			byte[] buffer = new byte[8192];
			int length;

			while ((length = input.read(buffer, 0, 8192)) > 0)
				output.write(buffer, 0, length);

			output.close();
			input.close();
		}
		catch (IOException e)
		{
			e.printStackTrace();
			return null;
		}

		return dbPath.getAbsolutePath();
	}

	public static EventDatabase getInstance()
	{
		return instance;
	}
}

File: android/app/src/main/java/com/tisza/esemenynaptar/database/EventDatabase.kt
package com.tisza.esemenynaptar.database

import android.content.*
import android.util.*
import androidx.room.*
import androidx.room.migration.*
import androidx.sqlite.db.*
import java.io.*

private const val EVENT_DATABASE_NAME = "event.db"
private const val EVENT_DATABASE_SRC_NAME = "event_src.db"
private const val VERSION = 2

lateinit var eventDatabase: EventDatabase
    private set

@Database(entities = [Event::class], version = VERSION, exportSchema = false)
@TypeConverters(CategoryConverter::class)
abstract class EventDatabase : RoomDatabase() {
    abstract fun eventDao(): EventDao
}

@Synchronized
fun initEventDatabase(context: Context) {
    val migrations = Array(VERSION - 1) { i ->
        UpdateDatabaseFromAssetsMigration(context, i + 1, i + 2)
    }

    eventDatabase = Room.databaseBuilder(context.applicationContext, EventDatabase::class.java, EVENT_DATABASE_NAME)
            .addMigrations(*migrations)
            .addCallback(EventDatabaseCallback(context))
            .build()
}

private class EventDatabaseCallback(val context: Context) : RoomDatabase.Callback() {
    override fun onCreate(db: SupportSQLiteDatabase) {
        updateDatabaseFromAsset(context, db)
    }
}

private class UpdateDatabaseFromAssetsMigration(val context: Context, startVersion: Int, endVersion: Int): Migration(startVersion, endVersion) {
    override fun migrate(db: SupportSQLiteDatabase) {
        updateDatabaseFromAsset(context, db)
    }
}

private fun updateDatabaseFromAsset(context: Context, db: SupportSQLiteDatabase) {
    db.setTransactionSuccessful()
    db.endTransaction()

    db.execSQL("ATTACH DATABASE ? AS event_src;", arrayOf<Any?>(copyAssetDatabaseToData(context)))

    db.beginTransaction()

    db.execSQL("DROP TABLE IF EXISTS event_old;")
    db.execSQL("CREATE TABLE event_old AS SELECT * FROM event;")
    db.execSQL("DELETE FROM event;")
    db.execSQL("INSERT INTO event SELECT event_src.event.year, event_src.event.month, event_src.event.day, event_src.event.category, event_src.event.text, ifnull(event_old.isLiked, 0) "
            + "FROM event_src.event LEFT JOIN event_old "
            + "ON event_src.event.year = event_old.year AND event_src.event.month = event_old.month AND event_src.event.day = event_old.day AND event_src.event.category = event_old.category;")
    db.execSQL("DROP TABLE IF EXISTS event_old;")
}

private fun copyAssetDatabaseToData(context: Context): String? {
    val dbPath = context.getDatabasePath(EVENT_DATABASE_SRC_NAME)
    dbPath.parentFile.mkdirs()
    try {
        val input = context.assets.open(EVENT_DATABASE_NAME)
        val output = FileOutputStream(dbPath)
        val buffer = ByteArray(8192)
        var length: Int
        while (input.read(buffer, 0, 8192).also { length = it } > 0)
            output.write(buffer, 0, length)
        output.close()
        input.close()
    } catch (e: IOException) {
        e.printStackTrace()
        return null
    }
    return dbPath.absolutePath
}
